File: Database/services/ventaProducto.py
from models.modelos import VentaProducto as VentaProductoModel
from schemas.ventas import VentaProducto
from typing import List

class VentaProductoService:
    
    def __init__(self, db) -> None:
        self.db = db

    # Obtener todos los productos de todas las ventas
    def get_venta_productos(self):
        result = self.db.query(VentaProductoModel).all()
        return result

    # Obtener los productos de una venta específica por ID de venta
    def get_venta_productos_by_venta_id(self, id_venta: int):
        result = self.db.query(VentaProductoModel).filter(VentaProductoModel.id_venta == id_venta).all()
        return result
    
    # Obtener un producto específico de una venta por ID de venta_producto
    def get_venta_producto_by_id(self, id_venta_producto: int):
        result = self.db.query(VentaProductoModel).filter(VentaProductoModel.id_venta_producto == id_venta_producto).first()
        return result
    
    # Crear un nuevo producto asociado a una venta
    def create_venta_producto(self, venta_producto: VentaProducto):
        # Procesar el id_categoria si es necesario, pero no lo guardamos en la base de datos
        if venta_producto.id_categoria:
            print(f"Procesando categoría: {venta_producto.id_categoria}")
        
        # Crear el objeto de la base de datos sin id_categoria
        new_venta_producto = VentaProductoModel(
            id_venta=venta_producto.id_venta,  # Asegúrate de pasar este campo correctamente
            id_producto=venta_producto.id_producto,
            cantidad=venta_producto.cantidad,
            subtotal=venta_producto.subtotal
        )
        self.db.add(new_venta_producto)
        self.db.commit()
        self.db.refresh(new_venta_producto)
        return new_venta_producto

    # Actualizar un producto asociado a una venta
    def update_venta_producto(self, id_venta_producto: int, data: VentaProducto):
        venta_producto = self.db.query(VentaProductoModel).filter(VentaProductoModel.id_venta_producto == id_venta_producto).first()
        if not venta_producto:
            return None

        # Actualizar los detalles del producto en la venta
        venta_producto.id_venta = data.id_venta
        venta_producto.id_producto = data.id_producto
        venta_producto.cantidad = data.cantidad
        venta_producto.subtotal = data.subtotal

        self.db.commit()
        return venta_producto

    # Eliminar un producto específico de una venta
    def delete_venta_producto(self, id_venta_producto: int):
        venta_producto = self.db.query(VentaProductoModel).filter(VentaProductoModel.id_venta_producto == id_venta_producto).first()
        if not venta_producto:
            return None
        
        self.db.delete(venta_producto)
        self.db.commit()
        return venta_producto

File: PHP/controllers/ventaProducto.php
<?php
include '../config.php'; // Conexión a la base de datos

// Obtener el método HTTP
$method = $_SERVER['REQUEST_METHOD'];

// Lógica CRUD de VentaProducto
if ($method === 'GET') {
    if (isset($_GET['id_venta_producto'])) {
        // Obtener un producto específico de una venta por ID de venta_producto
        $id_venta_producto = intval($_GET['id_venta_producto']);
        $query = "SELECT * FROM venta_producto WHERE id_venta_producto = ?";
        $stmt = $conn->prepare($query);
        $stmt->execute([$id_venta_producto]);
        $venta_producto = $stmt->fetch(PDO::FETCH_ASSOC);

        echo json_encode($venta_producto);
    } elseif (isset($_GET['id_venta'])) {
        // Obtener todos los productos de una venta específica
        $id_venta = intval($_GET['id_venta']);
        $query = "SELECT * FROM venta_producto WHERE id_venta = ?";
        $stmt = $conn->prepare($query);
        $stmt->execute([$id_venta]);
        $venta_productos = $stmt->fetchAll(PDO::FETCH_ASSOC);

        echo json_encode($venta_productos);
    } else {
        // Obtener todos los productos de todas las ventas
        $query = "SELECT * FROM venta_producto";
        $stmt = $conn->query($query);
        $venta_productos = $stmt->fetchAll(PDO::FETCH_ASSOC);

        echo json_encode($venta_productos);
    }
} elseif ($method === 'POST') {
    // Crear un nuevo producto asociado a una venta
    $data = json_decode(file_get_contents("php://input"), true);

    // Procesar los datos para la inserción
    $query = "INSERT INTO venta_producto (id_venta, id_producto, cantidad, subtotal) VALUES (?, ?, ?, ?)";
    $stmt = $conn->prepare($query);
    $stmt->execute([
        $data['id_venta'],
        $data['id_producto'],
        $data['cantidad'],
        $data['subtotal']
    ]);

    echo json_encode(["message" => "Producto asociado a la venta creado exitosamente"]);
} elseif ($method === 'PUT') {
    // Actualizar un producto específico de una venta
    $data = json_decode(file_get_contents("php://input"), true);
    $id_venta_producto = intval($data['id_venta_producto']);

    // Actualizar los detalles del producto
    $query = "UPDATE venta_producto SET id_venta = ?, id_producto = ?, cantidad = ?, subtotal = ? WHERE id_venta_producto = ?";
    $stmt = $conn->prepare($query);
    $stmt->execute([
        $data['id_venta'],
        $data['id_producto'],
        $data['cantidad'],
        $data['subtotal'],
        $id_venta_producto
    ]);

    echo json_encode(["message" => "Producto asociado a la venta actualizado exitosamente"]);
} elseif ($method === 'DELETE') {
    // Eliminar un producto específico de una venta
    $id_venta_producto = intval($_GET['id_venta_producto']);

    // Eliminar el registro de la base de datos
    $query = "DELETE FROM venta_producto WHERE id_venta_producto = ?";
    $stmt = $conn->prepare($query);
    $stmt->execute([$id_venta_producto]);

    echo json_encode(["message" => "Producto asociado a la venta eliminado exitosamente"]);
} else {
    // Método no permitido
    http_response_code(405);
    echo json_encode(["message" => "Método no permitido"]);
}
?>
